### PR TITLE
Add tikv write stall blog

### PR DESCRIPTION
**Pull request by dcalvin**
Add blog for publishing on Nov 12  "How to Troubleshoot RocksDB Write Stalls in TiKV"

- [ ] I have added necessary CTA buttons
- [ ] I have resized the images to <= 200 KB
- [ ] I have included necessary SEO keywords in filenames

### CODE REVIEW

File: how-to-troubleshoot-rocksdb-write-stalls-in-tikv.md [hunk 1]
```diff
@@ -0,0 +1,180 @@
...
+date: 2021-11-12
+summary: This troubleshooting guide discusses write performance degradation in TiKV related to the Rocks DB built-in write stall feature.
+tags: ['TiKV']
+categories: ['Engineering']
+image: /images/blog/troubleshoot-rocksdb-write-stalls-in-tikv.png
+---
+
+**Author:** Mike Barlow (Solutions Architect at PingCAP)
+
+**Editor:** Jose Espinoza, Tina Yang, Tom Dewan
```
Comment (ran-huang) [inline how-to-troubleshoot-rocksdb-write-stalls-in-tikv.md:13]:
```suggestion
**Editors:** Jose Espinoza, Tina Yang, Tom Dewan
```

File: how-to-troubleshoot-rocksdb-write-stalls-in-tikv.md [hunk 2]
```diff
@@ -0,0 +1,180 @@
+---
+title: How to Troubleshoot RocksDB Write Stalls in TiKV
+author: ['Mike Barlow']
+date: 2021-11-12
+summary: This troubleshooting guide discusses write performance degradation in TiKV related to the Rocks DB built-in write stall feature.
+tags: ['TiKV']
```
Comment (ran-huang) [inline how-to-troubleshoot-rocksdb-write-stalls-in-tikv.md:6]:
```suggestion
tags: ['TiKV', 'Tutorial']
```

File: how-to-troubleshoot-rocksdb-write-stalls-in-tikv.md [hunk 3]
```diff
@@ -0,0 +1,180 @@
...
+
+There are background tasks (jobs) that are specifically responsible for flushing memtables to disk and compacting SST files. By default `max-background-jobs` is set to 8 or CPU cores to -1, whichever is smaller. 
+
+There can be many memtables. Remember TiDB has a minimum of one memtable for each column family (default, write, lock, and raft). As mentioned before, when the memtable reaches the `write-buffer-size` limit it becomes read-only and is flagged to be flushed to disk as an SST file. There are a limited number of tasks that flush memtables to disk, which can be identified with `max-background-flushes` limit. The default is 2 or max-background-jobs / 4, whichever is bigger. 
+
+
+> **NOTE**
```
Comment (ran-huang) [inline how-to-troubleshoot-rocksdb-write-stalls-in-tikv.md:80]:
```suggestion
> **NOTE:**
```

File: how-to-troubleshoot-rocksdb-write-stalls-in-tikv.md [hunk 4]
```diff
@@ -0,0 +1,180 @@
...
+
+    For TiDB 5.0, the default is 8.
+* `[rocksdb|raftdb].max-background-flushes` 
+
+    The default value is set to 2 or `max_background_jobs` / 4, whichever is bigger.
+
+Grafana graphs:
+
+![Immutable memtable number](media/immutable-memtable-number.png)
+<div class="caption-center">Immutable memtable number</div> _
```
Comment (ran-huang) [inline how-to-troubleshoot-rocksdb-write-stalls-in-tikv.md:104]:
```suggestion
<div class="caption-center">Immutable memtable number</div>
```

File: how-to-troubleshoot-rocksdb-write-stalls-in-tikv.md [hunk 5]
```diff
@@ -0,0 +1,180 @@
...
+By default, when the count of level 0 SST files reaches 20, RocksDB will slow writes for the entire database. If the count of level 0 SST files reaches 36, RocksDB will stall writes for the entire database.
+
+Compaction is triggered when the number of level 0 SST files for a column family reaches 4. However, compaction may not occur. Parallel compactions on different column families can occur simultaneously.
+
+Configuration parameters:
+
+* `[rocksdb|raftdb].[defaultcf|writecf|lockcf].level0-file-num-compaction-trigger` 
+
+    The default is 4.
+* `[rocksdb|raftdb].[defaultcf|writecf|lockcf].level0-slowdown-writes-trigger`**`
```
Comment (ran-huang) [inline how-to-troubleshoot-rocksdb-write-stalls-in-tikv.md:128]: ?

**Review by ran-huang — COMMENTED** [review]
Please fix CI failure.

File: how-to-troubleshoot-rocksdb-write-stalls-in-tikv.md (hunk 5 above)
Comment (dcalvin) [reply how-to-troubleshoot-rocksdb-write-stalls-in-tikv.md]: fixed.

File: how-to-troubleshoot-rocksdb-write-stalls-in-tikv.md [hunk 6]
```diff
@@ -0,0 +1,177 @@
...
+![Write stall duration](media/write-stall-duration.png)
+<div class="caption-center">Write stall duration</div>
+
+With this knowledge, we can dig deeper into write stalls.
+
+RocksDB will trigger write stalls if there are too many:
+
+* [Memtables](#bookmark=id.hlmn1aylq4qf)
+* [Level 0 SST (Sorted String Table) files](#bookmark=id.cye5mbth2h5x) 
+* [Pending compaction bytes](#bookmark=id.j7eeqy4rw57t)
```
Comment (ran-huang) [inline how-to-troubleshoot-rocksdb-write-stalls-in-tikv.md:55]:
```suggestion
* [Memtables](#too-many-memtables)
* [Level 0 SST (Sorted String Table) files](#too-many-level-0-sst-files) 
* [Pending compaction bytes](#too-many-pending-compaction-bytes)
```

File: how-to-troubleshoot-rocksdb-write-stalls-in-tikv.md [hunk 7]
```diff
@@ -0,0 +1,177 @@
...
+
+![How to Troubleshoot RocksDB Write Stalls in TiKV](media/troubleshoot-rocksdb-write-stalls-in-tikv.png)
+
+[TiDB](https://pingcap.com/products/tidb), an open-source, distributed NewSQL database, can experience write performance degradation for several reasons. This troubleshooting guide discusses write performance degradation related to the [Rocks DB](https://docs.pingcap.com/tidb/stable/rocksdb-overview) built-in write stall feature. RocksDB is an open source, mature, and high performance key-value store. It is optimized for fast, low latency storage such as flash drives and high-speed disk drives.
+
+We will also discuss how to resolve this issue in [TiKV](https://tikv.org/), a highly scalable, low latency, and easy to use key-value database that [uses RocksDB as its storage engine](https://pingcap.com/blog/rocksdb-in-tikv).
+
+When RocksDB can't flush and compact data promptly, it uses a feature called "stalls" to try and slow the amount of data coming into the engine.  Write stalls include pausing all writes and limiting the number of writes.
+
+From [Github](https://github.com/facebook/rocksdb/wiki/Write-Stalls):
```
Comment (lilin90) [inline how-to-troubleshoot-rocksdb-write-stalls-in-tikv.md:23]:
```suggestion
From [GitHub](https://github.com/facebook/rocksdb/wiki/Write-Stalls):
```

File: how-to-troubleshoot-rocksdb-write-stalls-in-tikv.md [hunk 8]
```diff
@@ -0,0 +1,177 @@
...
+
+TiKV has two instances of RocksDB (RaftDB and KVdb). RaftDB has only one column family. KVDB has three column families (default, write, and lock).
+
+## Write stall user experiences 
+
+Because TiKV depends on RocksDB, when the application or process does a large number of writes, TiDB response time (latency) can degrade significantly. This latency seems to affect all writes and not just writes to a specific table.
+
+There are a couple of Grafana graphs that can help identify whether RocksDB's write stall is the cause and needs additional investigation. 
+
+We can use the following Grafana charts located in the section
```
Comment (lilin90) [inline how-to-troubleshoot-rocksdb-write-stalls-in-tikv.md:40]:
```suggestion
We can use the following Grafana charts located in the section:
```

File: how-to-troubleshoot-rocksdb-write-stalls-in-tikv.md [hunk 9]
```diff
@@ -0,0 +1,177 @@
...
+     The default 128 MB. If you increase this parameter, the possibility of an OOM also increases.
+* `[rocksdb|raftdb].[defaultcf|writecf|lockcf]max-write-buffer-number` 
+
+    The default is 5. If you increase this parameter, the possibility of an OOM also increases
+* `[rocksdb|raftdb].max-background-jobs` 
+
+    For TiDB 5.0, the default is 8.
+* `[rocksdb|raftdb].max-background-flushes` 
+
+    The default value is set to 2 or `max_background_jobs` / 4, whichever is bigger.
```
Comment (ran-huang) [inline how-to-troubleshoot-rocksdb-write-stalls-in-tikv.md:95]:
```suggestion

* `[rocksdb|raftdb].[defaultcf|writecf|lockcf]max-write-buffer-number` 

    The default is 5. If you increase this parameter, the possibility of an OOM also increases.

* `[rocksdb|raftdb].max-background-jobs` 

    For TiDB 5.0, the default is 8.

* `[rocksdb|raftdb].max-background-flushes` 

    The default value is set to 2 or `max_background_jobs` / 4, whichever is bigger.
```

File: how-to-troubleshoot-rocksdb-write-stalls-in-tikv.md [hunk 10]
```diff
@@ -0,0 +1,177 @@
...
+
+There are background tasks (jobs) that are specifically responsible for flushing memtables to disk and compacting SST files. By default `max-background-jobs` is set to 8 or CPU cores to -1, whichever is smaller. 
+
+There can be many memtables. Remember TiDB has a minimum of one memtable for each column family (default, write, lock, and raft). As mentioned before, when the memtable reaches the `write-buffer-size` limit it becomes read-only and is flagged to be flushed to disk as an SST file. There are a limited number of tasks that flush memtables to disk, which can be identified with `max-background-flushes` limit. The default is 2 or max-background-jobs / 4, whichever is bigger. 
+
+> **NOTE:**
+>
+>Prior to TiKV v4.0.9, `max-background-flushes` is not available, and the number of background flush tasks is set to `max-background-jobs` / 4.
```
Comment (lilin90) [inline how-to-troubleshoot-rocksdb-write-stalls-in-tikv.md:78]:
```suggestion
> Prior to TiKV v4.0.9, `max-background-flushes` is not available, and the number of background flush tasks is set to `max-background-jobs` / 4. 
```

File: how-to-troubleshoot-rocksdb-write-stalls-in-tikv.md [hunk 11]
```diff
@@ -0,0 +1,177 @@
...
+
+By default, when the count of level 0 SST files reaches 20, RocksDB will slow writes for the entire database. If the count of level 0 SST files reaches 36, RocksDB will stall writes for the entire database.
+
+Compaction is triggered when the number of level 0 SST files for a column family reaches 4. However, compaction may not occur. Parallel compactions on different column families can occur simultaneously.
+
+Configuration parameters:
+
+* `[rocksdb|raftdb].[defaultcf|writecf|lockcf].level0-file-num-compaction-trigger` 
+
+    The default is 4.
```
Comment (ran-huang) [inline how-to-troubleshoot-rocksdb-write-stalls-in-tikv.md:123]:
```suggestion
    The default is 4.
    
```

File: how-to-troubleshoot-rocksdb-write-stalls-in-tikv.md [hunk 12]
```diff
@@ -0,0 +1,177 @@
...
+* `[rocksdb|raftdb].[defaultcf|writecf|lockcf].level0-slowdown-writes-trigger`
+
+    The default is 20.
```
Comment (ran-huang) [inline how-to-troubleshoot-rocksdb-write-stalls-in-tikv.md:126]:
```suggestion
    The default is 20.
    
```

File: how-to-troubleshoot-rocksdb-write-stalls-in-tikv.md [hunk 13]
```diff
@@ -0,0 +1,177 @@
...
+For example, in the diagram below, the total pending compaction is 72 GB. If the soft-pending-compaction-bytes-limit is set to 64 GB (default), write stalls will be triggered.
+
+![Total pending compaction](media/total-pending-compaction.png)
+<div class="caption-center">Total pending compaction</div>
+
+Configuration parameters:
+
+* `[rocksdb|raftdb].soft-pending-compaction-bytes-limit` 
+
+    The default is 64 GB.
```
Comment (ran-huang) [inline how-to-troubleshoot-rocksdb-write-stalls-in-tikv.md:155]:
```suggestion
    The default is 64 GB.
    
```

File: how-to-troubleshoot-rocksdb-write-stalls-in-tikv.md [hunk 14]
```diff
@@ -0,0 +1,183 @@
...
+    The default is 36.
+
+Grafana graphs:
+
+![Level 0 SST file number](media/level0-sst-file-number.png)
+<div class="caption-center">Level 0 SST file number</div>
+
+### Too many pending compaction bytes
+
+A Log-Structured Merge-Tree (LSM) consists of multiple levels. Each level can include zero or more SST files. Level 0 is identified as a "higher hevel" and each incrementing level (level 1, level 2, etc.) is identified as a lower level. (See the diagram below.)
```
Comment (lilin90) [inline how-to-troubleshoot-rocksdb-write-stalls-in-tikv.md:143]:
```suggestion
A Log-Structured Merge-Tree (LSM) consists of multiple levels. Each level can include zero or more SST files. Level 0 is identified as a "higher level" and each incrementing level (level 1, level 2, etc.) is identified as a lower level. (See the diagram below.)
```

File: how-to-troubleshoot-rocksdb-write-stalls-in-tikv.md [hunk 15]
```diff
@@ -0,0 +1,183 @@
...
+image: /images/blog/troubleshoot-rocksdb-write-stalls-in-tikv.png
+---
+
+**Author:** Mike Barlow (Solutions Architect at PingCAP)
+
+**Editors:** Jose Espinoza, Tina Yang, Tom Dewan
+
+![How to Troubleshoot RocksDB Write Stalls in TiKV](media/troubleshoot-rocksdb-write-stalls-in-tikv.png)
+
+[TiDB](https://pingcap.com/products/tidb), an open-source, distributed NewSQL database, can experience write performance degradation for several reasons. This troubleshooting guide discusses write performance degradation related to the [Rocks DB](https://docs.pingcap.com/tidb/stable/rocksdb-overview) built-in write stall feature. RocksDB is an open source, mature, and high performance key-value store. It is optimized for fast, low latency storage such as flash drives and high-speed disk drives.
```
Comment (dcalvin) [inline how-to-troubleshoot-rocksdb-write-stalls-in-tikv.md:17]:
```suggestion
[TiDB](https://pingcap.com/products/tidb), an open-source, distributed NewSQL database, can experience write performance degradation for several reasons. This troubleshooting guide discusses write performance degradation related to the [Rocks DB](https://docs.pingcap.com/tidb/stable/rocksdb-overview) built-in write stall feature. RocksDB is an open-source, mature, and high-performance key-value store. It is optimized for fast, low latency storage such as flash drives and high-speed disk drives.
```

File: how-to-troubleshoot-rocksdb-write-stalls-in-tikv.md [hunk 16]
```diff
@@ -0,0 +1,183 @@
+---
+title: How to Troubleshoot RocksDB Write Stalls in TiKV
+author: ['Mike Barlow']
+date: 2021-11-12
+summary: This troubleshooting guide discusses write performance degradation in TiKV related to the Rocks DB built-in write stall feature.
```
Comment (CaitinChen) [inline how-to-troubleshoot-rocksdb-write-stalls-in-tikv.md:5]:
```suggestion
summary: This troubleshooting guide discusses write performance degradation in TiKV related to the RocksDB built-in write stall feature.
```

File: how-to-troubleshoot-rocksdb-write-stalls-in-tikv.md (hunk 15 above)
Comment (CaitinChen) [inline how-to-troubleshoot-rocksdb-write-stalls-in-tikv.md:17]:
```suggestion
[TiDB](https://pingcap.com/products/tidb), an open-source, distributed NewSQL database, can experience write performance degradation for several reasons. This troubleshooting guide discusses write performance degradation related to the [RocksDB](https://docs.pingcap.com/tidb/stable/rocksdb-overview) built-in write stall feature. RocksDB is an open-source, mature, and high performance key-value store. It is optimized for fast, low latency storage such as flash drives and high-speed disk drives.
```

**Review by ran-huang — APPROVED** [review]
LGTM

**Pull request merged**